doc/neps/nep-00XX-extensible-dtypes.rst.md

```diff
@@ -0,0 +1,414 @@
+       * DTypes which are not flexible, must have a `default_descr` for
+         that canonical implementation which should be a singleton.
+       * Flexible dtypes must define some additional attributes and
+         additional slots on `CastingImpl`.
+   * byteorder (not strictly necessary)
+   * is_native
+       * Instead of byteorder, we may want an `is_native` flag (we could just reuse the ISNBO flag – "is native byte order"), this flag signals that the data is stored in the default/canonical way. In practice this is always an NBO check, but generalization should be possible. A use case would be to have a complex-conjugated instance of Complex which is not the normal representation and typically has to be cast).
+   * alignment information for structured dtype construction. (May be possible to guess if the itemsize is fixed, at least for some dtypes, but maybe better to force to be provided).
+   * ...
+
+**Methods currently defined by ``ArrFuncs``** (due to the opaque nature of this struct, read access to it will remain supported at least for builtin dtypes for some time; It may be useful to detect and error on changes of the struct).
+   * sorting, argsorting
+   * take
+   * byteswap (``copyswapn``)
+   * etc.
+
+Where some of the last slots should be redefined as (generalized) ufuncs
+and deprecated.
+Note that the metadata described in point 2 describes a specfic instance,
+while methods/slots live on the dtype class.
+
+### C level implementation
+
+The C-level implementation of dtypes should largely follow similar designs
+as CPython. Unlike CPython, however, the layout shall be completely hidden
+from the downstream user, with most methods not accessable (initially).
+
+Thus, to create a new data type from the C-level, it will be necessary to
+use a code similar to this (modelled after [PEP-384]):
+
+```C
+static struct PyArrayDtypeMethodDef abstract_methods[] = {
+    {npy_adt_cast_from_to, cast_from_to},
+    {npy_adt_cast_to_from, cast_to_from},
+    {0, NULL}
+}
+
+
+static struct PyArrayDtypeMethodDef concrete_methods[] = {
+    {npy_cdt_take, take},
+    // TODO: Decide how to handle the multiple sorting methods:
+    {npy_cdt_sort, sort},
+    {0, NULL}
+}
+
+// ALTERNATIVE: Could use strings, instead of npy_*_... integers/enum.
+
+typedef struct {
+    /* Possibly PyTypeSpec (to simplify creation) */
+    /* Other necessary information */
+    int flags,  // (some flags, such as flexible yes/no.)
+    *PyArrayDtypeMethodDef abstract_methods,
+    *PyArrayDTypeMethodDef concrete_methods,
+} PyArrayDTypeSpec
+
+ 
+ /*
+  * Note that PyArrayDescr_Type will be typed as a DtypeMeta struct.
+  * Users may extend it. The function below initializes the DTypeMeta class.
+  * The legacy registration will create a new DTypeMeta intance
+  * (dtype subclass) and initalize it based on the existing information.
+  */
+PyObject* PyArrayDescrType_InitFromSpec(PyArrayDescr_Type *user_dtype, PyArrayDTypeSpec *);
```

Comment:
GH's rst-Rendering is broken below this line due to uneven number of asterisks (which makes things a bit hard to read)

Reply:
There is also the problem that I started on hackmd with markdown, but wrote some things more as rst, should convert it soon to just be real rst and not use hackmd